clippy_utils/src/check_proc_macro.rs

```diff
@@ -541,6 +541,102 @@ fn ident_search_pat(ident: Ident) -> (Pat, Pat) {
+            let (start, _) = qpath_search_pat(&path);
+            (start, Pat::Str("}"))
+        },
+        rustc_hir::PatKind::TupleStruct(path, _, _) => {
+            let (start, _) = qpath_search_pat(&path);
+            (start, Pat::Str(")"))
+        },
+        rustc_hir::PatKind::Or(plist) => {
+            // documented invariant
+            debug_assert!(plist.len() >= 2);
+            let (start, _) = pat_search_pat(tcx, plist.first().unwrap());
+            let (_, end) = pat_search_pat(tcx, plist.last().unwrap());
+            (start, end)
+        },
+        rustc_hir::PatKind::Never => (Pat::Str("!"), Pat::Str("")),
+        rustc_hir::PatKind::Tuple(_, _) => (Pat::Str("("), Pat::Str(")")),
+        rustc_hir::PatKind::Box(p) => {
+            let (_, end) = pat_search_pat(tcx, p);
+            (Pat::Str("box"), end)
+        },
+        rustc_hir::PatKind::Deref(_) => (Pat::Str("deref!("), Pat::Str(")")),
+        rustc_hir::PatKind::Ref(p, _) => {
+            let (_, end) = pat_search_pat(tcx, p);
+            (Pat::Str("&"), end)
+        },
+        rustc_hir::PatKind::Expr(expr) => pat_search_pat_expr_kind(expr),
+        rustc_hir::PatKind::Guard(pat, guard) => {
+            let (start, _) = pat_search_pat(tcx, pat);
+            let (_, end) = expr_search_pat(tcx, guard);
+            (start, end)
+        },
+        rustc_hir::PatKind::Range(None, None, range) => match range {
+            rustc_hir::RangeEnd::Included => (Pat::Str("..="), Pat::Str("")),
+            rustc_hir::RangeEnd::Excluded => (Pat::Str(".."), Pat::Str("")),
+        },
+        rustc_hir::PatKind::Range(r_start, r_end, range) => {
+            let (start, _) = match r_start {
+                Some(e) => pat_search_pat_expr_kind(e),
+                None => match range {
+                    rustc_hir::RangeEnd::Included => (Pat::Str("..="), Pat::Str("")),
+                    rustc_hir::RangeEnd::Excluded => (Pat::Str(".."), Pat::Str("")),
+                },
+            };
+
+            let (_, end) = match r_end {
+                Some(e) => pat_search_pat_expr_kind(e),
+                None => match range {
+                    rustc_hir::RangeEnd::Included => (Pat::Str(""), Pat::Str("..=")),
+                    rustc_hir::RangeEnd::Excluded => (Pat::Str(""), Pat::Str("..")),
+                },
+            };
+            (start, end)
+        },
+        rustc_hir::PatKind::Slice(_, _, _) => (Pat::Str("["), Pat::Str("]")),
+    }
+}
+
+fn pat_search_pat_expr_kind(expr: &crate::PatExpr<'_>) -> (Pat, Pat) {
+    match expr.kind {
+        crate::PatExprKind::Lit { lit, negated } => {
+            let (start, end) = lit_search_pat(&lit.node);
+            if negated { (Pat::Str("!"), end) } else { (start, end) }
+        },
+        crate::PatExprKind::ConstBlock(_block) => (Pat::Str("const {"), Pat::Str("}")),
+        crate::PatExprKind::Path(path) => qpath_search_pat(&path),
+    }
+}
```

Comment:
You can add imports for `rustc_hir::{PatKind, PatExpr, PatExprKind}`. `Pat` I understand conflicts so is fine to leave prefixed